Api Docs

arielsvn opened the pull request:
## Issue Number

#687 

## Purpose/Implementation Notes

Adding docs for the API, some views have to be refactored so that swagger can pick up all metadata.

- [x] Refactor experiments endpoint to use [ListApiView](https://www.django-rest-framework.org/api-guide/generic-views/#listapiview)
- [x] Remove old search views
- [x] Remove `/compendia` endpoint and use `/computed_files?is_compendia=True` instead. Also update serializer to avoid sending S3 url
- [x] Move `POST /dataset/create` to `POST /dataset` to make it restful

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests



## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

cgreene commented:
When this gets deployed, @arielsvn can you ping me on slack to follow up with a contracting firm who may be helping us to build an API client? I think this will help them get their estimate together.

arielsvn commented:
Removed the WIP, this should be ready for review.

![image](https://user-images.githubusercontent.com/1882507/59210992-cc38f080-8b7c-11e9-92a9-d0282e0711c2.png)

Seems `main_tests` are failing due to an error downloading files from ncbi, will try to re-run them later.

wvauclain commented on api/data_refinery_api/tests.py:
```diff
@@ -342,320 +328,42 @@ def test_compendia(self):
         self.assertEqual(3, len(response_json))
         # Prove that the download_url field is missing and not None.
-        self.assertEqual('NotPresent', response_json[0].get('download_url', 'NotPresent'))
+        self.assertEqual(None, response_json[0].get('download_url', None))
```
Why did you make this change?

arielsvn replied:
good catch! I'll update the serializers and restore this.

wvauclain commented on api/data_refinery_api/tests.py:
```diff
@@ -342,320 +328,42 @@ def test_compendia(self):
         self.assertEqual(3, len(response_json))
         self.assertIsNone(response_json[0]['download_url'])
 
-    def test_search_and_filter(self):
```
Why was this test removed?

arielsvn replied:
We changed the search and started using elastic search. There are tests for that here https://github.com/AlexsLemonade/refinebio/blob/c67e006cc524ad9f57119d5a7f96f0f9a333f6bd/api/data_refinery_api/tests.py#L864

Also I wasn't able to make this one work with elastic search, seems to be tricky.

wvauclain commented:
Hitting `/docs/` gave me this error:

```
Internal Server Error: /docs/
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 124, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 483, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 443, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 480, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/schemas/views.py", line 31, in get
    schema = self.schema_generator.get_schema(request, self.public)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/schemas/generators.py", line 286, in get_schema
    links = self.get_links(None if public else request)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/schemas/generators.py", line 324, in get_links
    link = view.schema.get_link(path, method, base_url=self.url)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/schemas/inspectors.py", line 195, in get_link
    fields += self.get_serializer_fields(path, method)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/schemas/inspectors.py", line 315, in get_serializer_fields
    serializer = view.get_serializer()
  File "/usr/local/lib/python3.6/site-packages/rest_framework/generics.py", line 111, in get_serializer
    kwargs['context'] = self.get_serializer_context()
  File "/home/user/data_refinery_api/views.py", line 416, in get_serializer_context
    token_id = self.request.META.get('HTTP_API_KEY', None)
AttributeError: 'NoneType' object has no attribute 'META'
```

wvauclain commented on api/data_refinery_api/views.py:
```diff
@@ -905,205 +733,138 @@ def get(self, request, format=None):
+    This can return valid S3 urls if a valid [token](#tag/token) is sent in the header `HTTP_API_KEY`.
     """
+    queryset = ComputationalResult.public_objects.all()
+    serializer_class = ComputationalResultSerializer
```
If I'm not mistaken, you don't need to set `serializer_class` if you're using `get_serializer_class()`

```
Typically, you must either set this attribute, or override the get_serializer_class() method.
```

kurtwheeler commented on common/data_refinery_common/migrations/0021_update_qn_bucket.py:
```diff
@@ -44,7 +44,7 @@ def update_qn_bucket(apps, schema_editor):
 
     dependencies = [
-        ('data_refinery_common', '0019_sample_is_blacklisted'),
+        ('data_refinery_common', '0020_auto_20190607_1505'),
```
Wait no, don't change this. This migration has already been run. Change the migration you're adding in this PR to have this migration as a dependency. (Also your dependency should be number 21, not number 20. This migrations is already number 20).

arielsvn commented:
@wvauclain @kurtwheeler thanks for reviewing, I just addressed your comments. Let me know if you see any other issues.

> Hitting /docs/ gave me this error:

Removed this endpoint, since we'll have swagger docs now.
